solutions/usaco-946.mdx

```diff
@@ -31,62 +48,81 @@ const ll factor2 = 2019201949LL;
  * each other
  */
 ll calc_length(ll a, ll b) {
-  a++, b++;
-  return (a * factor1 + b * factor2) % M;
+	a++, b++;
+	return (a * factor1 + b * factor2) % M;
 }
 
 /**
  * Prim's Algorithm for dense graph to build the MST by scanning
  * @return the edges used to construct the MST
  */
 vector<ll> prim(int N) {
-  vector<ll> dist(N, M);
-  vector<bool> visited(N, false);
-
-  for (int i = 0; i < N; i++) {
-	// find the nearest node to the current MST
-	int min_dist_node = -1;
-	for (int j = 0; j < N; j++) {
-	  if (!visited[j] && (min_dist_node < 0 || dist[j] < dist[min_dist_node])) {
-		min_dist_node = j;
-	  }
-	}
+	vector<ll> dist(N, M);
+	vector<bool> visited(N, false);
+
+	for (int i = 0; i < N; i++) {
+		// find the nearest node to the current MST
+		int min_dist_node = -1;
+		for (int j = 0; j < N; j++) {
+			if (!visited[j] && (min_dist_node < 0 || dist[j] < dist[min_dist_node])) {
+				min_dist_node = j;
+			}
+		}
 
-	// add this node to the MST
-	visited[min_dist_node] = true;
-
-	// update the distances between the MST and other unconnected nodes
-	for (int j = 0; j < N; j++) {
-	  if (!visited[j]) {
-		dist[j] = min(
-			dist[j], 
-			calc_length(min(min_dist_node, j), max(min_dist_node, j))
-		);
-	  }
+		// add this node to the MST
+		visited[min_dist_node] = true;
+
+		// update the distances between the MST and other unconnected nodes
+		for (int j = 0; j < N; j++) {
+			if (!visited[j]) {
+				dist[j] = min(
+					dist[j], 
+					calc_length(min(min_dist_node, j), max(min_dist_node, j))
+				);
+			}
+		}
 	}
-  }
-  return dist;
+	return dist;
 }
 
 int main() {
-  freopen("walk.in", "r", stdin);
-  freopen("walk.out", "w", stdout);
+	freopen("walk.in", "r", stdin);
+	freopen("walk.out", "w", stdout);
 
-  int N, K;
-  cin >> N >> K;
+	int N, K;
+	cin >> N >> K;
 
-  vector<ll> mst = prim(N);
-  sort(mst.begin(), mst.end());
-  cout << mst[N - K] << endl;
+	vector<ll> mst = prim(N);
+	sort(mst.begin(), mst.end());
+	cout << mst[N - K] << endl;
 }
 ```
 </CPPSection>
 
 </LanguageSection>
 
+
 ## Solution 2: Kruskal's Algorithm & Two-Pass Radix Sort 
 
-Kruskal with two iterations of counting sort passes in 1966 ms ...
+Alternatively, we can also use Kruskal's Algorithm to construct our MST. Note
+that only the step of sorting all edges takes too long time. If we manage to
+sort all edges in linear time, we might be able to solve it also with Kruskal's
+Algorithm as well.
+
+To sort these edges, we can apply a two-pass radix sort. In particular, we first
+use counting sort to sort the first 16 bits ($[0, 2^16)$) of the weights. Then,
+in the second iteration, we sort the rest 16 bits ($[2^16, 2^32)$). After the
+sorting process, we apply the unmodified Kruskal's Algorithm on the edges and
+$M$ is equal to the weight of the $N - K + 1$-th longest edge of the MST.
+
+Time Complexity: $\mathcal{O}(N^2)$
+
+
+<Warning title="TLE">
+Some large test cases might get TLE due to the bad constant factor for the 
+radix sort. 
```

Comment:
so the below solutions only _sometimes_ time out right

Reply:
Benq asserts that it only *sometimes* timeout (and he has got ACC in 19~~ms), but both his and my code got TLE on the last three test cases yesterday when I tested it. 

though the official analysis *does* mention this as a solution alternative